src/variety/railpool.js

```diff
@@ -0,0 +1,659 @@
+	}
+})(["railpool"], {
+	//---------------------------------------------------------
+	// マウス入力系
+	MouseEvent: {
+		inputModes: {
+			edit: ["border", "empty", "clear", "info-line"],
+			play: ["line", "peke", "info-line"]
+		},
+		mouseinput_auto: function() {
+			if (this.puzzle.playmode) {
+				if (this.btn === "left") {
+					if (this.mousestart || this.mousemove) {
+						this.inputLine();
+					} else if (this.mouseend && this.notInputted()) {
+						this.inputpeke();
+					}
+				} else if (this.btn === "right") {
+					if (this.mousestart || this.mousemove) {
+						this.inputpeke();
+					}
+				}
+			} else if (this.puzzle.editmode) {
+				if (this.btn === "left") {
+					if (this.mousestart || this.mousemove) {
+						this.inputborder();
+					} else if (this.mouseend && this.notInputted()) {
+						this.setcursor(this.getcell());
+					}
+				} else if (this.btn === "right") {
+					this.inputempty();
+				}
+			}
+		},
+		// taken from simpleloop
+		mouseinput_other: function() {
+			if (this.inputMode === "empty") {
+				this.inputempty();
+			}
+		},
+		inputempty: function() {
+			var cell = this.getcell();
+			if (cell.isnull || cell === this.mouseCell) {
+				return;
+			}
+			if (this.inputData === null) {
+				this.inputData = cell.isEmpty() ? 0 : 7;
+			}
+
+			cell.setQues(this.inputData);
+			cell.drawaround();
+			this.mouseCell = cell;
+		}
+	},
+
+	//---------------------------------------------------------
+	// キーボード入力系
+	KeyEvent: {
+		enablemake: true,
+
+		keyinput: function(ca) {
+			this.key_inputqnum_railpool(ca);
+		},
+		// taken from lohkous
+		key_inputqnum_railpool: function(ca) {
+			var cell = this.cursor.getc(),
+				nums = cell.qnums,
+				val = [];
+
+			if (("1" <= ca && ca <= "9") || ca === "-") {
+				var num = ca !== "-" ? +ca : -2;
+				var clear = false;
+				if (this.prev === cell) {
+					for (var i = 0; i < nums.length; i++) {
+						if (num === -2 || num !== nums[i]) {
+							val.push(nums[i]);
+						} else {
+							clear = true;
+						}
+					}
+				}
+				if (!clear) {
+					if (nums.length < 4) {
+						val.push(num);
+					} else {
+						val = [num];
+					}
+				}
+			} else if (ca === "BS") {
+				if (nums.length > 1) {
+					for (var i = 0; i < nums.length - 1; i++) {
+						val.push(nums[i]);
+					}
+				}
+			} else if (ca === " ") {
+				val = [];
+			} else {
+				return;
+			}
+
+			cell.setNums(val);
+
+			this.prev = cell;
+			cell.draw();
+		}
+	},
+
+	// (almost) everything between here and Border is taken from tapaloop.js
+	//---------------------------------------------------------
+	// 盤面管理系
+	Cell: {
+		qnums: [],
+		noLP: function(dir) {
+			return this.isEmpty();
+		},
+		setNums: function(val) {
+			this.setQnums(val);
+			this.setQans(0);
+			this.setQsub(0);
+		},
+		setQnums: function(val) {
+			if (this.puzzle.pzpr.util.sameArray(this.qnums, val)) {
+				return;
+			}
+			this.addOpeQnums(this.qnums, val); // i don't know what this is for
```

Comment:
I also don't know why both setQnums and addOpeQnums do a check for no-ops. It seems okay to remove `addOpeQnums` and replace all references to a call to `setQnums`. Adding the ObjectOperation2 should be done in this function.

Reply:
That comment doesn't only refer to that line: i don't know what Operations are, what's the `klass`, and what's the goal of `ObjectOperation2`. `addOpeQnums` is only called in line 194; without understanding the code, it doesn't seem equivalent to calling `setQnums`, since the latter will always call the `addOpeQnums` functionality and the former does only so depending on the parameter `isrec` (whose origin & function I ignore). Given that the code is the same in tapa, tapaloop, and lohkous, I'm happy leaving it as-is, even if it would be okay to simplify it.

Reply:
Alright, it's fine to not merge the functions, as long as this comment is removed.

Operation subclasses are used for the undo stack, and the call to `addOpeQnums` contains the current value and the new value of qnums. Since having multiple qnums in one cell isn't part of the standard framework, Tapa includes a subclass `ObjectOperation2` specifically for replacing the entire list of numbers. This was then copied to Tapa-Like Loop and Lohkous.